recipes/dspsr/meta.yaml

```diff
@@ -0,0 +1,93 @@
+{% set name = "dspsr" %}
+{% set version = "2022-02-17" %}
+{% set sha256 = "835246c81d78f25b4620cbabf50dfa7ac3a7c1e3145046102a3bb8a54bc5e57b" %}
+
+package:
+    name: "{{ name | lower }}"
+    version: "{{ version | replace('-','.') }}"
+
+source:
+    url: https://sourceforge.net/projects/{{ name }}/files/{{ name }}/{{ version }}/{{ name }}-{{ version }}.tar.gz
+    sha256: {{ sha256 }}
+    patches:
+        - macos_fmemopen.patch  # [osx]
+        - python_ldflags.patch
+        - shell_for_flags.patch
+        
+build:
+    number: 0
+    skip: true  # [win or (python_impl == 'pypy')]
+    run_exports:
+      - {{ pin_subpackage('dspsr', exact=True) }}
+
+requirements:
+    build:
+        - {{ compiler('cxx') }}
+        - {{ compiler('c') }}
+        - {{ compiler('fortran') }}  # [osx]
```

Comment:
Yes I completely agree it's strange, this is what I was trying to explain in an earlier comment.  All of these packages (including tempo2) are indirect dependencies of dspsr, via psrchive (except fftw which is used directly by dspsr, but is also a dependency of psrchive).  When I remove them from the dspsr requirements list I get overlinking (_not_ overdepending) errors for OSX.  When I include them, I get overdepending warnings for Linux.  Hence the current situation where they are marked as OSX-only requirements.

I'm wondering if this is related to the removal of the `-dead_strip_dylibs` option in build.sh.  In the past I found this was required in order for the python API to work on OSX.  I could try keeping this option in and see what happens.  Other suggestions welcome of course!

Reply:
OK. My understanding is that you have a dependency tree like (simplified version):
```
dspsr -> psrchive, fftw, pgplot
psrchive -> fftw, tempo2, gsl, libcblas
pgplot -> libpng, zlib
```
Thus, your meta.yaml should be like this:
```yaml
host:
  - psrchive
  - fftw
  - pgplot
run:
  - {{ pin_compatible('psrchive', exact=true) }}  # because psarchive recipe is missing run_export
  - {{ pin_compatible('pgplot') }}  # because pgplot recipe missing run_export
  - fftw  # fftw recipe has run_export so adding explicitly is not necessary
  # libpng, zlib, tempo2, gsl, libcblas should NOT be listed here because they are not direct dependencies of dspsr
```

If you get an overlinking error from conda-build with the meta above, that means conda has inspected your binaries and found that one of them links DIRECTLY to a transitive dependency (you call it indirect dependency). For example, it finds that dspsr links directly to tempo2. This is an error because if psrchive was to remove tempo2 from its dependencies, then people newly installing dspsr would experience errors because tempo2 would no longer be in the dependency tree but the dspsr binary links to it.

As you have observed, adding these packages to the dependencies of dspsr will silence the overlinking error, BUT this should only be the solution if dspsr ACTUALLY directly depends on the package. For example, in the case of FFTW, you say that dspsr needs the headers of FFTW to compile, so it is a direct dependency. It is possible to tell a compiler to dynamically link binaries that don't need each other. The proper solution for overlinking errors for gsl, tempo2, etc it to patch/fix your build to prevent them from being linked when unnecessary.

From a quick google search, it looks like reenabling the `-dead_strip_dylibs` compiler option is the path we want to investigate. We want the binaries of dspsr to only link to libraries that they are actually using. If `-dead_strip_dylibs` doesn't work then you will need to manually edit the build process to remove unused linking declarations.

Reply:
Right, this is starting to make a bit more sense.  Reenabling `dead_strip_dylibs` does seem to help with dependencies but still breaks the Python interface.  There are unresolved symbol errors upon testing `import dspsr`:
```
2022-03-01T21:06:31.1303590Z import: 'dspsr'
2022-03-01T21:06:34.7150430Z Traceback (most recent call last):
2022-03-01T21:06:34.7252090Z   File "/Users/runner/mambaforge/conda-bld/dspsr_1646167814904/test_tmp/run_test.py", line 2, in <module>
2022-03-01T21:06:34.7352380Z     import dspsr
2022-03-01T21:06:34.7455820Z   File "/Users/runner/mambaforge/conda-bld/dspsr_1646167814904/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.7/site-packages/dspsr.py", line 15, in <module>
2022-03-01T21:06:34.7553550Z     import _dspsr
2022-03-01T21:06:34.7656970Z ImportError: dlopen(/Users/runner/mambaforge/conda-bld/dspsr_1646167814904/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.7/site-packages/_dspsr.so, 2): Symbol not found: __ZN3dsp8SKLimits11calc_limitsEv
2022-03-01T21:06:34.7757140Z   Referenced from: /Users/runner/mambaforge/conda-bld/dspsr_1646167814904/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.7/site-packages/../../libdspdsp.0.dylib
```
I think this is because the linker strips libraries that it thinks are unused, however we don't really know what is/isn't used until python tries to call it.

I'll poke at it a bit more.  If it comes to it, I think disabling the python API for OSX would be OK with me; do you see any issue with that?

Reply:
> I think this is because the linker strips libraries that it thinks are unused, however we don't really know what is/isn't used until python tries to call it.

Assuming you are a developer of dspsr, you *should* know what is/isn't used. 😆 

> I'll poke at it a bit more. If it comes to it, I think disabling the python API for OSX would be OK with me; do you see any issue with that?

I think that's fine. There's no requirement that packages provide feature parity on all platforms.

Reply:
OK, python is removed for OSX.  I also had to remove one binary (dmsmear) which also has linking errors when the library-stripping option is enabled.  I can't really tell why only this one should be affected, but it's a non-essential utility program so I just removed it.  Let me know if you think this needs any more changes.